FileService_GetFile_RabbitMq

lalagvana opened the pull request:
- Added Consumer, Request and Response in FileService for getting file through RabbitMQ.
- Configured MassTransit in FileService, ProjectService and UserService.
- Added Request and Response in ProjectService and UserService.
- Fixed some not really necessary stuff.

Now it works correctly, I suppose :)

lalagvana committed:
Get File from FileService for UserService and ProjectService

-Added IOperationResult interface in Kernel project.
-Added Mapper, Repository, Consumer, Request and Response in FileService for getting file through RabbitMQ.
-Configured MassTransit in FileService.
-Added Request, Controller in ProjectService and UserService.
-Fixed some not really necessary stuff.

MihailKulikov commented on DigitalOffice/Services/ProjectService/Controllers/ProjectController.cs:
```diff
@@ -16,6 +20,26 @@ public Project GetProjectInfoById([FromServices] IGetProjectInfoByIdCommand comm
+                Id = id
+            };
+
+            var a = await client.GetResponse<IOperationResult<object>>(request);
```
a?

mekk1t replied:
As I see it, this should mean an anonymous object. But Lena should have given it a proper name :)

lalagvana replied:
Fixed

MihailKulikov commented on DigitalOffice/Services/UserService/Controllers/UserController.cs:
```diff
@@ -21,5 +25,25 @@ public bool CreateUser([FromServices] IUserCreateCommand command, [FromBody] Use
+                Id = id
+            };
+
+            var a = await client.GetResponse<IOperationResult<object>>(request);
```
а?

lalagvana replied:
Fixed

djino97 commented on DigitalOffice/Services/FileService/Broker/Consumers/GetFileConsumer.cs:
```diff
@@ -0,0 +1,35 @@
+
+            var response = mapper.Map(
+                repository.GetFile(request.Id));
+
+            await context.RespondAsync(response);
+        }
```
Use IOperationResult

lalagvana replied:
Fixed

djino97 commented on DigitalOffice/Services/FileService/Broker/Consumers/GetFileConsumer.cs:
```diff
@@ -0,0 +1,35 @@
+        {
+            var request = context.Message;
+
+            var response = mapper.Map(
+                repository.GetFile(request.Id));
```
May be move this logic to the command class? Single Responsibility Principle.

djino97 commented on DigitalOffice/Services/UserService/UserService.csproj:
```diff
@@ -7,16 +7,25 @@
+    <PackageReference Include="RabbitMQ.Client" Version="6.1.0" />
   </ItemGroup>
 
   <ItemGroup>
     <Folder Include="Models" />
     <Folder Include="Controllers\" />
   </ItemGroup>
```
This is not good.

lalagvana replied:
Fixed

ifelseelif commented on DigitalOffice/Services/ProjectService/Startup.cs:
```diff
@@ -29,43 +32,61 @@ public Startup(IConfiguration configuration)
+                {
+                    cfg.Host("localhost", h =>
+                    {
+                        h.Username("username");
```
Change to something more informative

lalagvana replied:
Fixed

ifelseelif commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -33,13 +35,34 @@ public void ConfigureServices(IServiceCollection services)
+                {
+                    cfg.Host("localhost", h =>
+                    {
+                        h.Username("username");
```
Change to something more informative

lalagvana replied:
Fixed

ifelseelif commented on DigitalOffice/Services/ProjectService/Program.cs:
```diff
@@ -12,6 +13,10 @@ public static void Main(string[] args)
             Host.CreateDefaultBuilder(args)
+                .ConfigureLogging(log =>
+                {
+                    log.AddConsole();
```
What is this for

lalagvana replied:
To add the Console logging provider.

ifelseelif commented on DigitalOffice/Services/ProjectService/Controllers/ProjectController.cs:
```diff
@@ -16,6 +20,26 @@ public Project GetProjectInfoById([FromServices] IGetProjectInfoByIdCommand comm
+            [FromQuery] Guid id,
+            [FromServices] IRequestClient<InternalGetFileRequest> client)
+        {
+            var request = new
```
Move this logic to a command

lalagvana committed:
Merge branch 'develop' into feature/FileService_GetFile_RabbitMq

djino97 commented on DigitalOffice/Services/FileService/Broker/Responses/InternalFileResponse.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace LT.DigitalOffice.Broker.Responses
+{
+    public class InternalFileResponse
+    {
+        public string Content { get; }
+        public string Extension { get; }
+        public string Name { get; }
+    }
```
Use interface, instead class

lalagvana replied:
Fixed

lalagvana committed:
Fixed references to Kernel

lalagvana committed:
Fixes

- Changed name 'a' to 'response'.
- Chenged to interfaces and changed name of IGetFileRequest and IGetFileResponse.
- Deleted extra folders in UserService.csproj.

KostinP commented on DigitalOffice/Kernel/IOperationResult.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Kernel
+{
+    public interface IOperationResult<T>
```
Add XML.

KostinP replied:
Why is this marked as resolved?

lalagvana replied:
Hmm, github doesn't catch some local changes. I'll fix it

lalagvana replied:
Now it's fixed

KostinP commented on DigitalOffice/Services/FileService/Broker/Requests/IGetFileRequest.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace LT.DigitalOffice.Broker.Requests
+{
+    public interface IGetFileRequest
```
Add XML.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Services/FileService/Broker/Responses/InternalFileResponse.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace LT.DigitalOffice.Broker.Responses
+{
+    public interface IFileResponse
```
Add XML.

KostinP commented on DigitalOffice/Services/ProjectService/Broker/Requests/IGetFileRequest.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace LT.DigitalOffice.Broker.Requests
+{
+    public interface IGetFileRequest
```
Add XML.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Services/ProjectService/Controllers/ProjectController.cs:
```diff
@@ -16,6 +20,26 @@ public Project GetProjectInfoById([FromServices] IGetProjectInfoByIdCommand comm
         }
 
+        [Route("getProjectFileById")]
+        [HttpGet]
```
Can be combined.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Services/UserService/Broker/Requests/IGetFileRequest.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace LT.DigitalOffice.Broker.Requests
+{
+    public interface IGetFileRequest
```
Add XML.

KostinP commented on DigitalOffice/Services/UserService/Controllers/UserController.cs:
```diff
@@ -21,5 +25,25 @@ public bool CreateUser([FromServices] IUserCreateCommand command, [FromBody] Use
         }
+
+        [Route("getUserAvatarById")]
+        [HttpGet]
```
Can be combined.

lalagvana replied:
Fixed

lalagvana committed:
Merge branch 'develop' into feature/FileService_GetFile_RabbitMq

lalagvana committed:
Fixes after merging develop

lalagvana committed:
Fix route

lalagvana committed:
Merge remote-tracking branch 'origin/develop' into feature/FileService_GetFile_RabbitMq

lalagvana committed:
A lot of fixes....

MihailKulikov commented on DigitalOffice/Tests/FileServiceTests/Consumers/GetFileConsumerTests.cs:
```diff
@@ -0,0 +1,139 @@
+        public async Task ShouldConsumeAndPublishWhenRequestIsValidAndRepositoryReturnsSomething()
+        {
+            repositoryMock
+                .Setup(repository => repository.CheckIfUserHasRight(It.IsAny<ICheckIfUserHasRightRequest>()))
```
auf

djino97 commented on DigitalOffice/Services/FileService/Broker/Consumers/GetFileConsumer.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            try
+            {
+                await context.RespondAsync<IOperationResult<IGetFileResponse>>(new
+                {
+                    Body = repository.GetFileById(context.Message.Id),
+                    IsSuccess = true,
+                    Errors = new List<string>()
+                });
+            }
+            catch (Exception exception)
+            {
+                await context.RespondAsync<IOperationResult<IGetFileResponse>>(new
+                {
+                    Body = false,
+                    IsSuccess = false,
+                    Errors = new List<string> { exception.Message }
+                });
```
DRY, move await context.RespondAsync<IOperationResult> behind construction try{}catch{}.

djino97 commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -33,6 +36,32 @@ public void ConfigureServices(IServiceCollection services)
             });
             services.AddControllers();
 
+            services.AddMassTransit(configurator =>
+            {
+                configurator.AddConsumer<GetFileConsumer>();
+
+                configurator.UsingRabbitMq((context, factoryConfigurator) =>
+                {
+                    const string serviceInfoSection = "ServiceInfo";
+
+                    var serviceName = Configuration.GetSection(serviceInfoSection)["Name"];
+                    var serviceId = Configuration.GetSection(serviceInfoSection)["Id"];
+
+                    factoryConfigurator.Host("localhost", "/", hostConfigurator =>
+                    {
+                        hostConfigurator.Username($"{serviceName}_{serviceId}");
+                        hostConfigurator.Password(serviceId);
+                    });
+
+                    factoryConfigurator.ReceiveEndpoint("FileService", endpointConfigurator =>
+                    {
+                        endpointConfigurator.ConfigureConsumer<GetFileConsumer>(context);
+                    });
+                });
+            });
```
May be this code move to the method?

djino97 commented on DigitalOffice/Services/FileService/appsettings.json:
```diff
@@ -9,5 +9,9 @@
     "SQLConnectionString": "Server=localhost;Database=FileDB;Trusted_Connection=True;"
+  },
+  "ServiceInfo": {
+    "Name": "FileService",
+    "Id": "B59045FD-1267-4549-839A-2CD3376C7999",
   }
```
use localhost port for Id?

djino97 commented on DigitalOffice/Services/ProjectService/Controllers/ProjectController.cs:
```diff
@@ -15,6 +19,25 @@ public Project GetProjectInfoById([FromServices] IGetProjectInfoByIdCommand comm
+            };
+
+            var response = await client.GetResponse<IOperationResult<object>>(request);
+            if (!response.Message.IsSuccess)
+            {
+                throw new Exception();
+            }
```
Must be exceptions processing for response message from broker. And this all logic must be move in execute method, there is call only Execute method.

djino97 commented on DigitalOffice/Services/ProjectService/Startup.cs:
```diff
@@ -35,30 +38,51 @@ public void ConfigureServices(IServiceCollection services)
+                configurator.UsingRabbitMq((context, factoryConfigurator) =>
+                {
+                    factoryConfigurator.Host("localhost", hostConfigurator =>
+                    {
+                        hostConfigurator.Username($"{serviceName}_{serviceId}");
+                        hostConfigurator.Password($"{serviceId}");
+                    });
```
Simple serviceId.

djino97 commented on DigitalOffice/Tests/FileServiceTests/Consumers/GetFileConsumerTests.cs:
```diff
@@ -0,0 +1,139 @@
+            consumerTestHarness = harness.Consumer(() => new GetFileConsumer(repositoryMock.Object));
+        }
+
+        [Test]
+        public async Task ShouldConsumeAndPublishWhenRequestIsValidAndRepositoryReturnsSomething()
+        {
```
ReturnsSomething?)

MihailKulikov replied:
NUnit supports only async test methods that return Task or Task<T>

djino97 commented on DigitalOffice/Tests/FileServiceTests/Consumers/GetFileConsumerTests.cs:
```diff
@@ -0,0 +1,139 @@
+
+                Assert.That(response.Message.Body, Is.False);
+                Assert.That(response.Message.IsSuccess, Is.False);
+                Assert.That(response.Message.Errors, Is.EquivalentTo(new List<string> { exceptionMessage }));
```
May be use AreEquals?

lalagvana committed:
Merge remote-tracking branch 'origin/develop' into feature/FileService_GetFile_RabbitMq

lalagvana committed:
Merge remote-tracking branch 'origin/develop' into feature/FileService_GetFile_RabbitMq

lalagvana committed:
Merging

lalagvana committed:
Merge branch 'develop' into feature/FileService_GetFile_RabbitMq

lalagvana committed:
Refactoring my code in FileService

lalagvana committed:
Small fixes

lalagvana committed:
Merge branch 'feature/FileService_GetFile_RabbitMq' of https://github.com/LT-Students/DigitalOffice into feature/FileService_GetFile_RabbitMq

lalagvana committed:
Now it works!!!!!

KostinP commented on DigitalOffice/Services/FileService/Repositories/Interfaces/IFileRepository.cs:
```diff
@@ -15,7 +15,6 @@ public interface IFileRepository
         /// <param name="file">File to add.</param>
         /// <returns>Id of the added file.</returns>
         Guid AddNewFile(DbFile file);
-
```
Why?

lalagvana replied:
Fixed

lalagvana committed:
Added response in ProjectService + small fix

lalagvana committed:
Added masstransit logic in ProjectService and UserService connected with GetFile

lalagvana committed:
Deleted not needed files

lalagvana committed:
Some small fixes

ifelseelif commented on DigitalOffice/Services/UserService/Broker/Responses/IGetFileResponse.cs:
```diff
@@ -0,0 +1,12 @@
+    /// </summary>
+    public interface IGetFileResponse
+    {
+        public string Content { get; }
```
I think, public not need

lalagvana replied:
Fixed

djino97 commented on DigitalOffice/Services/FileService/Startup.cs:
```diff
@@ -38,6 +43,35 @@ public void ConfigureServices(IServiceCollection services)
+                        host.Password(serviceId);
+                    });
+
+                    cfg.ReceiveEndpoint($"{serviceName}", ep =>
```
Simple serviceName, without interpolation

lalagvana replied:
Fixed

djino97 commented on DigitalOffice/Services/FileService/Broker/Responses/IGetFileResponse.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    /// <summary>
+    /// DTO for dispatch file information through a message broker.
+    /// </summary>
+    public interface IGetFileResponse
+    {
```
IFileResponse, without Get

lalagvana replied:
Fixed

djino97 commented on DigitalOffice/Tests/FileServiceTests/Consumers/GetFileConsumerTests.cs:
```diff
@@ -0,0 +1,117 @@
+{
+    internal class GetFileConsumerTests
+    {
+        private static byte[] contentByte = Convert.FromBase64String("RGlnaXRhbCBPZmA5Y2U=");
+        private readonly string contentString = Convert.ToBase64String(contentByte);
+        private readonly string extension = ".jpg";
+        private readonly string name = "File";
```
Why not transfer initialization to setup  method?

lalagvana replied:
Fixed

lalagvana committed:
Small fixes

lalagvana committed:
Merge remote-tracking branch 'origin/develop' into feature/FileService_GetFile_RabbitMq

KostinP commented on DigitalOffice/Services/ProjectService/Broker/Requests/IGetFileRequest.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace LT.DigitalOffice.Broker.Requests
+{
+
```
Remove extra empty line

KostinP commented on DigitalOffice/Services/UserService/Broker/Requests/IGetFileRequest.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace LT.DigitalOffice.Broker.Requests
+{
+
```
Remove extra empty line.

KostinP reviewed (approved):
Ok!

lalagvana committed:
Removed some spaces

lalagvana committed:
Merge branch 'develop' into feature/FileService_GetFile_RabbitMq

lalagvana committed:
Merge branch 'develop' into feature/FileService_GetFile_RabbitMq

mekk1t commented on Services/FileService/Startup.cs:
```diff
@@ -38,6 +43,35 @@ public void ConfigureServices(IServiceCollection services)
             ConfigureRepositories(services);
             ConfigureValidators(services);
+            ConfigureRabbitMq(services);
+
+            services.AddMassTransitHostedService();
+        }
```
This can be moved to ConfigureRabbitMq

lalagvana replied:
fixed

mekk1t commented on Services/FileService/Startup.cs:
```diff
@@ -38,6 +43,35 @@ public void ConfigureServices(IServiceCollection services)
+
+            services.AddMassTransitHostedService();
+        }
+
+        private void ConfigureRabbitMq(IServiceCollection services)
+        {
```
I think this is wrong naming. I use "ConfigureMassTransit" and this should be the correct name since we really do configure MassTransit, don't we?

lalagvana replied:
fixed

mekk1t commented on Services/FileService/Broker/Requests/IGetFileRequest.cs:
```diff
@@ -0,0 +1,12 @@
+    /// Represents request for GetFileConsumer in MassTransit logic.
+    /// </summary>
+    public interface IGetFileRequest
+    {
+        Guid FileId { get; }
+    }
```
In UserService and ProjectService this field is named "Id". No conflicts happened?

lalagvana replied:
fixed

mekk1t reviewed (approved):
Check the fields names in IGetFileRequest, because they differ.
Also have you checked this manually? Via Postman and 3 simultaneous projects started up? If that's working, then you can merge.

lalagvana committed:
Small fixes
